crates/findex/src/pyo3/py_api.rs

```diff
@@ -554,3 +232,188 @@ impl InternalFindex {
+            &label.0,
+        );
+        let rt = pyo3_unwrap!(
+            tokio::runtime::Runtime::new(),
+            "async runtime creation error"
+        );
+        pyo3_unwrap!(rt.block_on(future), "error blocking for upsert");
+        Ok(())
+    }
+
+    /// Recursively search Findex graphs for `Location` corresponding to the
+    /// given `Keyword`.
+    ///
+    /// Parameters
+    ///
+    /// - `keywords`                : keywords to search using Findex
+    /// - `token`                   : Findex token
+    /// - `label`                   : public label used in keyword hashing
+    /// - `max_results_per_keyword` : maximum number of results to fetch per
+    ///   keyword
+    /// - `max_depth`               : maximum recursion level allowed
+    /// - `fetch_chains_batch_size` : batch size during fetch chain
+    ///
+    /// Returns: List[IndexedValue]
+    #[staticmethod]
+    #[pyo3(signature = (
+        keywords,
+        token,
+        label,
+        max_result_per_keyword = 4_294_967_295,
+        max_depth = 100,
+        fetch_chains_batch_size = 0,
+        base_url = None
+    ))]
+    #[allow(clippy::too_many_arguments)]
+    pub fn search(
+        keywords: Vec<ToKeyword>,
+        token: &str,
+        label: &LabelPy,
+        max_result_per_keyword: usize,
+        max_depth: usize,
+        fetch_chains_batch_size: usize,
+        base_url: Option<String>,
+    ) -> PyResult<HashMap<KeywordPy, Vec<LocationPy>>> {
+        let mut findex = pyo3_unwrap!(FindexCloudRust::new(token, base_url), "error reading token");
+        let master_key = findex.token.findex_master_key.clone();
+
+        let keywords_set: HashSet<Keyword> =
+            keywords.iter().map(|keyword| keyword.0.clone()).collect();
+
+        let rt = pyo3_unwrap!(
+            tokio::runtime::Runtime::new(),
+            "async runtime creation error"
+        );
+
+        let results = pyo3_unwrap!(
+            rt.block_on(
+                findex.search(
+                    &keywords_set,
+                    &master_key,
+                    &label.0,
+                    max_result_per_keyword,
+                    max_depth,
+                    NonZeroUsize::new(fetch_chains_batch_size)
+                        .unwrap_or(SECURE_FETCH_CHAINS_BATCH_SIZE),
+                    0,
+                )
+            ),
+            "error blocking for search"
+        );
+
+        Ok(search_results_to_python(results))
+    }
+
+    /// Generate a new Findex Cloud token with reduced permissions
+    #[staticmethod]
+    pub fn derive_new_token(token: String, search: bool, index: bool) -> PyResult<String> {
+        let mut token = pyo3_unwrap!(Token::from_str(&token), "error reading token");
+
+        pyo3_unwrap!(
+            token.reduce_permissions(search, index),
+            "error reducing token permissions"
+        );
+
+        Ok(token.to_string())
+    }
+
+    /// Generate a new random Findex Cloud token
```

Comment:
Missing parameter documentation
```suggestion
    /// Generates a new random Findex Cloud token.
```